### PR TITLE
Improve Window_Message timing accuracy - Messages 6 of N

### PR DESCRIPTION
**Pull request by fmatthew5876**
Depends on #1964 

Fixes more cases of #1706

This is yet another refactor of Window_Message. I've tried to clean up the Update loop to make it more understandable.

Unfortunately the number of edge cases here is just insane. So making this code readable is a real challenge.

There are a number of notable issues
* RPG_RT animates open/close after the main loop. Since we don't do this, I had to put in hacks to emulate the timing.
* I've finally nailed down the exact timing (and the rationale for why it works this way) for window open, close, and the interpreter blocking. 
* There are many, many subtle edge cases previously unknown with character timing. Even more-so with speed=1. Many have been discovered and added.

For this PR I have a much improved workflow to measure accuracy. To time any message, you can setup the following:

Foreground Event:
```
Msg: \^
FlashScreen 31,31,31,31 0.0s (no wait)
Var1 = 0
Msg: MESSAGE TO TEST
Msg: Time \v[1]
```

Parallel event
```
Var1+=1
```

To get individual character timings within a message, you have to do screen captures. I am doing this with `simplescreenrecorder` on linux by recording RPG_RT at 60fps. Then I use `ffmpeg -i vid.mkv frames%0004d.png` to dump the frames to png files and look at them.

The flash command in the above event helps identify exactly the frame where the message event command is executed and pushed into the messaging system.
The above construct will measure the exact number of frames the message takes to render.

Instead of writing page after page of test cases in issues. I have a test game with lots of message tests.

The test game is here. 

https://github.com/fmatthew5876/EasyRPG-InterpreterTestGame

All maps under the "Messages" Map tree have the #1706 test cases.
In particular the map `Messages / MESSAGE CRAZY` has a lot of NPCs with different text timing edge cases. Just talk to each NPC in RPG_RT and Player and compare.

#1706 Retested:
- [x] 1 - 16 Pass
- [x] 19 -24 Pass
- [x] 25 Timers retested - still working
- [x] 28-30 Pass
- [x] 31 We don't emulate system graphic corruption and shouldn't
- [x] 32-39 Pass

#1706 Newly fixed:
- [x] 26 Inn interruption tested with text, choices, and number input. 

- [x] 40 - tested and working
- [x] 41 - Timings match

#1706 Still broken:
- [ ] 17 & 18 Pushed to later PR
- [ ] 27 Test 1 - Pushed to later PR
- [ ] 27 Test 2 - box doesn't close.
- [ ] 42 - We don't interrupt the message
- [ ] 43 - Currently off by 1 frame.
- [ ] 44 - We set `show_message` chunk but it gets cleared every frame due to `Game_Map::UpdateForegroundEvents()` always calling `Game_Interpreter::Clear()` even when no events run.
- [ ] 45 Don't match

### CODE REVIEW

**Comment by fmatthew5876** [comment]
This one is ready. There are still some minor things left. I can address them in later PRs.

This has been tested and now text timings are very accurate.

**Comment by fmatthew5876** [comment]
This one just underwent rebase hell getting it over the stuff that just went in. Leaving a note for myself to do some testing of this again once dependencies are in and this is closer to being merged.

**Comment by fmatthew5876** [comment]
jenkins test this please

**Comment by carstene1ns** [comment]
Another test for this is PeuterisGrey. After the intro some textboxes are broken in current master.

**Comment by fmatthew5876** [comment]
After #2019 goes in I'll rebase this again and retest everything.

**Comment by carstene1ns** [comment]
This issue is fixed here (offset):
screenshot_3

This one is still present (name/item/variable substitution):
screenshot_0 screenshot_1 screenshot_2

**Comment by fmatthew5876** [comment]
Rebased. Still needs to be retested.

**Comment by fmatthew5876** [comment]
I reran my test cases and they still pass. I fixed the Peuteris Grey issue, that was a regression from when substitutions were refactored.

I've re-examined this complicated logic which animates the text box when transitions occur. The implementation I had was incomplete in several cases and not always frame accurate, so I've taken that out. It can be revisited in another PR.

**Comment by fmatthew5876** [comment]
Are we still planning to get this in for 6.2 or punt to next release?

If we're punting, the "Message \n[] respect actor renames" commit should be broken out into a separate PR and included in 6.2. That's a very simple fix for a regression.

**Comment by Ghabry** [comment]
I would prefer to only have the high priority commits in for 0.6.2 and the other commits for after.

**Comment by fmatthew5876** [comment]
> 
> 
> I would prefer to only have the high priority commits in for 0.6.2 and the other commits for post-release.

That's the only one. Everything else here is a big nasty refactor perfecting message frame timing.

**Comment by Ghabry** [comment]
This fixes a minor visual regression that didn't exist in 0.6.1: The Face graphic is not permanently shown. There is a 1 frame flicker everytime the page changes.

File: src/window_message.cpp [hunk 1]
```diff
@@ -243,95 +329,116 @@ void Window_Message::InsertNewLine() {
...
+	close_started_this_frame = false;
+	close_finished_this_frame = false;
 
-		if (!Game_Message::IsMessagePending() && IsVisible() && !closing) {
-			// Start the closing animation
-			SetCloseAnimation(Game_Battle::IsBattleRunning() ? 0 : message_animation_frames);
-			// This frame a foreground event may push a new message and interupt the close animation.
-			allow_next_message = true;
-		}
+#if 0
```
Comment (Ghabry) [inline src/window_message.cpp:381]: #if 0

**Comment by Ghabry** [comment]
Well with 0.6.2 out we can now get this here in and push to Android Beta (really Beta channel, not main xD) to get some feedback.

File: src/window_message.cpp (hunk 1 above)
Comment (fmatthew5876) [reply src/window_message.cpp]: Removed

**Comment by Ghabry** [comment]
Kinda message (choice) related:

There is a known regression about Inn continuation being broken when it costs money: https://github.com/rueter37/Player/commit/43a22be917b44837ea9fd9ba9d9aa8870b4239b1 (ugly solution)

The problem is that the async op is defined through a Choice and is never processed. Is this something that must be solved in messages, e.g. by passing in the async context:

```
Window_Message::InputChoice window_message.cpp:714
Window_Message::Update window_message.cpp:416
Game_Message::Update game_message.cpp:203
Game_Map::Update game_map.cpp:965 <- Has async ctx
```

Should thsi be solved on message or on interpreter side?

**Comment by fmatthew5876** [comment]
Are we sure this is still a bug? I've tested inn quite a bit and never saw such an issue.

Interpreter continuations have been gone for quite some time now. 

What happens now is the inn continuation gets stored in the `PendingMessage` which passes it to `Window_Message`. When you select your choice, the window_message calls the inn callback immediately, and this updates the interpreter subcommand path.

This works similarly for all other continuations such as name changes, battle results, shop buy/sell, etc..

**Comment by fmatthew5876** [comment]
The inn regression is completely fixed in #2212

The cross section of that PR with this one is minimal. I'm fine with rebaseding either one onto the other.

**Comment by Ghabry** [comment]
Inn is in(n) so can be merged.
When 0.6.2.1 is out the current master can be pushed to Android Beta channel.
I don't want to push the beta before because the internal release number for Android is based on "commits in repo" and it is possible that stable is rejected because the number is lower 🙄🤔
